invenio_rdm_records/services/components/pids.py

```diff
@@ -20,42 +20,48 @@
 class ExternalPIDsComponent(ServiceComponent):
     """Service component for pids."""
 
-    def _validate_pid(self, scheme, pid):
+    def _validate_pid(self, scheme, pid, record, provider=None):
         """Call provider to validate the given PID."""
-        provider_name = pid.get("provider")
-        client = pid.get("client")
-        provider = self.service.get_provider(scheme, provider_name, client)
-        if provider:
-            success, errors = provider.validate(**pid)
-            if errors:
-                raise ValidationError(message=errors, field_name="pids")
-
-    def _validate_pids(self, pids):
+        if not provider:  # In case we do not need to calculate it again
+            provider_name = pid.get("provider")
+            client = pid.get("client")
+            provider = self.service.get_provider(scheme, provider_name, client)
+            if not provider:
+                raise ValidationError(
+                    message=_(f"Provider {provider_name} not found for PID " +
+                              f"type {scheme}"),
+                    field_name="pids",
+                )
+
+        # NOTE: provider should not be None by now, if not configured should
+        # fail in `_validate_pid_schemes`
+        success, errors = provider.validate(record=record, **pid)
+        if errors:
+            raise ValidationError(message=errors, field_name=f"pids.{scheme}")
+
+    def _validate_pids(self, pids, record):
         """Validate an iterator of PIDs."""
-        verified_pids = set()
         pids_providers = self.service.config.pids_providers
         for scheme, providers in pids_providers.items():
-            if scheme not in pids:
-                continue
-
-            pid = pids.get(scheme, {})
-            self._validate_pid(scheme, pid)
-            verified_pids.add(scheme)
+            pid = pids.get(scheme)
+            if pid:
+                self._validate_pid(scheme, pid, record)
 
-        # ensure that there are no extra PIDs for which there is no config
+    def _validate_pid_schemes(self, pids):
+        """Validate the pid schemes are supported by the service."""
+        pids_providers = set(self.service.config.pids_providers.keys())
         all_pids = set(pids.keys())
```

Comment:
using the `pids.keys()` will give you the type e.g. `doi` that should match the `scheme`.... we are assuming this is a 1-1 match right? because of how the config is done... This might make it fail in many cases... (see tests)

Reply:
Yes, you are right. Honestly we probably have to rework the config at some point because it can lead to many inconsistencies.
Are you suggesting something in particular?

Reply:
I am out of ideas for this one... But I fully agree it need rewrite at config level to make a structure that actually works...